File: edp_java/src/main/Seat.java
package main;

public class Seat {
    private static final Seat[] SEATS = new Seat[Philosopher.list().length];
    static {
        for (int i = 0; i < SEATS.length; i++)
            SEATS[i] = new Seat();
    }

    public static Seat[] seats() {
        return SEATS.clone();
    }

    public static synchronized Seat availableSeat() {
        for (Seat seat : SEATS)
            if (!seat.occupied)
                return seat;
        return null;
    }

    private static int nextNumber = 0;
    private final int number;
    private boolean occupied = false;

    public Seat() {
        this.number = nextNumber++;
    }

    public int number() {
        return number;
    }

    public synchronized boolean isOccupied() {
        return occupied;
    }

    public synchronized boolean attemptToOccupy() {
        if (!occupied) {
            occupy();
            return true;
        }
        return false;
    }

    private synchronized void occupy() {
        occupied = true;
    }

    public synchronized void vacate() {
        System.out.println(String.format("Seat %d is vacated.", number));
        occupied = false;
    }
}

File: edp_go/seat.go
package main

import "sync"

type Seat struct {
	number   int
	occupied bool
	mutex    sync.Mutex
}

func (s *Seat) Number() int { return s.number }
func (s *Seat) IsOccupied() bool {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	return s.occupied
}
func (s *Seat) AttemptToOccupy() bool {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	if !s.occupied {
		s.occupied = true
		return true
	}
	return false
}
func (s *Seat) Vacate() {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	s.occupied = false
}

type SeatManager struct {
	seats []*Seat
}

var seatManager = &SeatManager{
	seats: make([]*Seat, len(PHILOSOPHER_NAMES)),
}

func init() {
	for i := range seatManager.seats {
		seatManager.seats[i] = &Seat{number: i}
	}
}

func (sm *SeatManager) Seats() []*Seat {
	result := make([]*Seat, len(sm.seats))
	copy(result, sm.seats)
	return result
}

func (sm *SeatManager) AvailableSeat() *Seat {
	for _, seat := range sm.seats {
		if !seat.IsOccupied() {
			return seat
		}
	}
	return nil
} 